### ISSUE TITLE
New subscribers receive no messages after plumtree_metadata_manager crashed

### ISSUE THREAD

**Issue by lancechentw** [vernemq/vernemq#715]
### Environment

- VerneMQ Version: 1.3.0
- OS: Container-Optimized OS on GKE
- Erlang/OTP version (if building from source):

### Expected behavior

- plumtree_metadata_manager should not crash
- New subscriber should receive messages if `plumtree_metadata_manager` has recovered

### Actual behaviour

- plumtree_metadata_manager crashed
- New subscribers does not receive any messages, while existing subscribers do receive messages

Hello guys, I believe I have ran into the same issue as reported in #547. I have a 4-node cluster where some nodes seem randomly stop publishing messages to new subscribers. However, existing subscribers seem working fine.

I have got three questions here
1. Should the case clause in https://github.com/erlio/plumtree/blob/master/src/plumtree_metadata_manager.erl#L472 handles an empty list?
2. Or should `plumtree_metadata_object:values(Metadata)` not return an empty list?
3. Does `plumtree_sup` respawn a new plumtree_metadata_manager? If so, why new subscribers fail to receive messages?

```
2018-05-22 07:10:06.360 [error] <0.239.0> CRASH REPORT Process plumtree_metadata_manager with 0 neighbours exited with reason: no case clause matching [] in plumtree_metadata_manager:store/2 line 472 in gen_server:terminate/7 line 812
2018-05-22 07:10:06.360 [error] <0.220.0> Supervisor plumtree_sup had child plumtree_metadata_manager started with plumtree_metadata_manager:start_link() at <0.239.0> exit with reason no case clause matching [] in plumtree_metadata_manager:store/2 line 472 in context child_terminated
2018-05-22 07:10:06.362 [error] <0.238.0> gen_server plumtree_broadcast terminated with reason: {{{case_clause,[]},[{plumtree_metadata_manager,store,2,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,472}]},{plumtree_metadata_manager,read_merge_write,2,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,457}]},{plumtree_metadata_manager,handle_call,3,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,316}]},{gen_server,try_handle_call,...},...]},...}
2018-05-22 07:10:06.363 [error] <0.238.0> CRASH REPORT Process plumtree_broadcast with 0 neighbours exited with reason: {{{case_clause,[]},[{plumtree_metadata_manager,store,2,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,472}]},{plumtree_metadata_manager,read_merge_write,2,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,457}]},{plumtree_metadata_manager,handle_call,3,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,316}]},{gen_server,try_handle_call,...},...]},...} in gen_server2:terminate/3 line 995
2018-05-22 07:10:06.363 [error] <0.220.0> Supervisor plumtree_sup had child plumtree_broadcast started with plumtree_broadcast:start_link() at <0.238.0> exit with reason {{{case_clause,[]},[{plumtree_metadata_manager,store,2,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,472}]},{plumtree_metadata_manager,read_merge_write,2,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,457}]},{plumtree_metadata_manager,handle_call,3,[{file,"/opt/vernemq/distdir/1.3.1/_build/default/lib/plumtree/src/plumtree_metadata_manager.erl"},{line,316}]},{gen_server,try_handle_call,...},...]},...} in context child_terminated
```

**Comment by dergraf** [comment]
@Lance0312 Hi Lance, thanks for reporting. The underlying `dvvset` is capable of returning an empty list, however we don't expect it to do so in this case... Something has to be wrong, I'll have to dig into the Plumtree Metadata code to see what it is. 

Plumtree Sup should restart those children. However, it is actually possible that the Routing Table subscriptions don't survive the restart. Will investigate this as well.

**Comment by larshesel** [comment]
hi @Lance0312 just curious, are you using any (custom) plugins or is it stock VerneMQ?

**Comment by Lance0312** [comment]
Hello @larshesel, only a `auth_on_register` webhook is enabled.

**Comment by Lance0312** [comment]
Also, `vmq_passwd` is enabled as well.

**Comment by Lance0312** [comment]
Hello @dergraf @larshesel, after a short period time of observation. It seems that the problem occurs much more frequently on a cluster of 4+ nodes (have tried 4, 5, and 6 nodes). I am now running the cluster on 3 nodes, though one of the nodes get killed by OOM and restart once a while.

Also, I have manually interpret the failed case clause, and found that it should not be resolved as an empty list. Do you have any ideas why it is resolved as an empty list?

Below is a last message which causes the crash
```
{
    merge,
    {
        {
            vmq,
            subscriber
        },
        {
            [],
            <<"mosq/Xc7Nx2nCRueH[cl^xB">>
        }
    },
    {
        metadata,
        {
            [
                {
                    'VerneMQ@broker-0.broker-cluster.default.svc.cluster.local',
                    5,
                    []
                },
                {
                    'VerneMQ@broker-1.broker-cluster.default.svc.cluster.local',
                    6,
                    []
                },
                {
                    'VerneMQ@broker-2.broker-cluster.default.svc.cluster.local',
                    7,
                    [
                        {
                            [
                                {
                                    'VerneMQ@broker-0.broker-cluster.default.svc.cluster.local',
                                    true,
                                    []
                                }
                            ],
                            {
                                1527,
                                949468,
                                491482
                            }
                        }
                    ]
                },
                {
                    'VerneMQ@broker-3.broker-cluster.default.svc.cluster.local',
                    2,
                    []
                }
            ],
            []
        }
    }
}
```

**Comment by ioolkos** [comment]
@Lance0312 thanks for details! is it always the same cluster node crashing?

Also, I'm going to mark this as 'bug', as we obviously need a solution for the empty list question.

**Comment by Lance0312** [comment]
@ioolkos, no, it is not always the same node crashing. Sometimes, several nodes crashed at the same time with the same last message (same subscriber field).

**Comment by larshesel** [comment]
Hi @Lance0312 I'm trying to investigate this a bit more. Do you know if the client `<<"mosq/Xc7Nx2nCRueH[cl^xB">>` was using `clean_session=false`? Are you perchance using `allow_multiple_sessions=on`?

**Comment by monsieurpug** [comment]
Hi, we are falling in the same issue.

**Comment by monsieurpug** [comment]
Falling into the same issue, on a cluster of three nodes, the reproduction steps are still unclear and we will try to provide more information on this topic. The version is VerneMQ 1.3.0. Enclosed the crash logs and the vernemq.conf.

[etc_vernemq_vernemq.21.txt](https://github.com/vernemq/vernemq/files/3032789/etc_vernemq_vernemq.21.txt)


[crash.log.0.txt](https://github.com/vernemq/vernemq/files/3032791/crash.log.0.txt)

**Comment by monsieurpug** [comment]
Is there any more data you would need to help checking into this issue ? :)

**Comment by ioolkos** [comment]
Hi @monsieurpug apologies for our delay in getting to this.
Is upgrading an option? If yes, we'd recommend that including switching to the new vmq_swc layer for distribution of metadata. Set `metadata_plugin = vmq_swc` in the vernemq.conf (unfortunately this will not take over the existing LevelDB databases, but create new ones).

**Comment by monsieurpug** [comment]
Thanks for the reply, upgrading might be an option, but I see that this feature is available from 1.7.0 but is still considered in Beta so to us that might be too early for production I guess.
Furthermore, what is the impact of not taking over existing LevelDB databases ? Do we lose existing databases ? Is a classic node-per-node upgrade from 1.3.0 in plumtree to 1.7.0 swc possible ?

**Comment by dergraf** [comment]
@monsieurpug I recommend to upgrade to a version `>=` VerneMQ 1.6. We introduced a new safe-guard that prevents the `plumtree_metadata_manager` from crashing in this rare case. Whenever VerneMQ would hit such a case we'd log a warning `Known issue, please report here https://github.com/erlio/vernemq/issues/715`  (which is this issue btw)

Regarding SWC: We currently don't have automatic migrations that transform the plumtree databases to the SWC databases. So you'd start with a fresh state.

**Comment by monsieurpug** [comment]
Thanks, that is good news. Do you know what are the consequences on the metadata when falling on the issue and not crashing thanks to the workaround ? Are there maybe inconsistencies between the nodes that could have consequences on the overall service like described here : https://github.com/vernemq/docker-vernemq/issues/67 ?

**Comment by dergraf** [comment]
The problems described in the mentioned issues should have been solved since 1.6.0 as the main problem causing the mentioned problems was the crashing of `plumtree_metadata_manager`. 

However, the root cause of course is and was the problematic metadata object, which at that point doesn't hold data anymore, but hasn't been properly deleted (that's our current hypothesis). So, our current approach is to re-delete the object to ensure it holds the proper tombstone value.

**Comment by monsieurpug** [comment]
Thanks for the clarification.
